❇️  사이드바 북마크 데이터 정합성 검사 기능 신규 개발

pureliture opened the pull request:
## 개요
이 Pull Request는 mirror 브랜치의 모든 변경사항을 메인 브랜치로 병합하기 위한 요청입니다. mirror 브랜치에는 북마크 데이터 관리, GitLab CI/CD 설정, 로깅 시스템 개선 등 다양한 기능 추가 및 개선사항이 포함되어 있습니다.
## 주요 변경사항
### 북마크 데이터 관련 기능
- 📝 북마크 데이터 YAML 구조 설계 문서 추가
- 🏷️ GitLab 관련 URL 정보를 북마크 데이터에 업데이트
- 🔥 중복 URL 항목 제거 및 주석 처리
- ✨ 북마크 YAML 검증 코드 개선

### CI/CD 및 파이프라인 개선
- 🔧 GitLab CI/CD 설정 파일 추가 및 업데이트
- 🗃️ Docker 레지스트리 인증을 위한 before_script 추가
- PipeLine Trigger Token 사용 리뷰 문서 추가

### 보안 및 로깅 개선
- 🔒 토큰 암호화/복호화 로직 개선
- 🔒 GitLab API 인증 로직 개선 (PAT 지원 추가)
- ✨ 로깅 시스템 적용 (`print` → `logger`로 변경)
- 📝 환경 변수 검증 및 에러 메시지 로깅 추가

### GitHub Actions 워크플로우
- 이슈에서 문서 생성을 위한 push-from-issue.yml 워크플로우 추가
- review-md.yml 워크플로우 추가 및 업데이트

### 문서화 개선
- GitLab Repo 구조 문서 작성 및 업데이트
- 다양한 리뷰 요청 문서 추가

## 테스트 결과
- 북마크 데이터 검증이 정상적으로 동작하는지 확인
- CI/CD 파이프라인이 정상 작동하는지 확인
- 토큰 암호화/복호화 기능이 예상대로 작동하는지 확인

## 리뷰 필요 사항
1. 북마크 데이터 YAML 구조가 적절한지 검토 필요
2. 암호화/복호화 로직의 보안성 확인 필요
3. 로깅 시스템 적용이 모든 부분에 일관되게 이루어졌는지 확인
4. GitLab CI/CD 설정이 프로젝트 요구사항에 맞게 구성되었는지 검토

pureliture committed:
Initial commit

(cherry picked from commit 04ff1ac76eceed6c70544a14bede54098be4f6f6)

pureliture committed:
:white_check_mark: Data Validator CI 설정 수정

- 특정 프로젝트 경로 조건 추가
- merge_request_event 규칙에 대한 처리 로직 업데이트

(cherry picked from commit 1240c213b35139a82542fcf9b70f897110519011)

pureliture committed:
🔧 GitLab CI 설정 업데이트

- Python 패키지 캐시 디렉토리 설정 추가
- 의존성 설치 전 pip 캐시 적용
- validate_bookmarks 스크립트 경로 수정
- 데이터 유효성 검사 스크립트 git clone 로직 추가

(cherry picked from commit c01c8c2e86db406d35935d0bef4794906bf307ff)

pureliture committed:
🚀 GitLab CI/CD 구성 및 Docker 지원 추가

- `.dockerignore` 및 `Dockerfile` 새로 추가
- GitLab CI/CD 파이프라인에 Docker 이미지 빌드 및 배포 단계 추가
- 기존 북마크 검증 작업 스크립트 경량화 및 Docker 이미지화

(cherry picked from commit f649206bf6948f2eab81daeccf166266344dca1c)

pureliture committed:
⚙️ GitLab CI 규칙 수정

- 파이프라인 소스 조건에 프로젝트 경로 검증 추가

(cherry picked from commit 24c774deedfa4451c0f1d8bfdcf7c68219348fa0)

pureliture committed:
📝 data-validator CI 구성 파일 분리 및 리팩토링

- docker-build-ci.yml 및 validation-ci.yml 신규 추가
- .gitlab-ci.yml의 구조 단순화 및 include 방식으로 변경

(cherry picked from commit 45174258c0a762f90ded146316f710fa43bd2cee)

pureliture committed:
merge conflict resolve

pureliture committed:
🔧 북마크 검증 및 데이터 처리 로직 개선

- 다중 프로젝트 북마크 검증 로직 및 CI/CD 환경 변수 설정 추가
- GitLab API를 활용한 프로젝트 및 파일 데이터 Fetch 기능 구현
- 북마크 데이터 및 YAML 검증 스크립트 리팩토링
- 배포 토큰 암호화를 위한 Token Manager 모듈 추가
- Docker 환경 설정 및 스크립트 통합 개선

(cherry picked from commit 5a999b036e31cf7396154f9147174addddbbe49b)

pureliture committed:
✨ 북마크 YAML 검증 코드 개선 및 CI 스크립트 업데이트

- 북마크 검증 코드의 주석 및 출력 메시지를 한글로 번역하여 가독성 향상
- CI 스크립트에 디버깅용 디렉토리 정보 및 YAML 파일 탐색 명령 추가
- 검증 로직에서 환경 변수 확인 및 오류 메시지 출력 기능 개선
- 불필요한 코드 제거 및 함수 호출 로직 간소화

(cherry picked from commit 4689c1d5162b087ee7a941c203317424a0d7d97b)

pureliture committed:
:zap: validate_bookmarks.py 환경 변수 검증 로직 개선

- 누락된 환경 변수를 출력하는 로직 추가
- 환경 변수 요구사항 안내 메시지 개선

(cherry picked from commit 9be8cb9e612f1954f71bd8b6dc7392fa86f68761)

pureliture committed:
🔒 GitLab API 인증 로직 개선

- 개인 액세스 토큰(PAT) 지원 추가
- 배포 토큰 및 PAT 기반 인증 헤더 생성 로직 통합
- 토큰 암호화 및 복호화 방식 코드 리팩토링
- 오류 및 상태 메시지를 개선하여 디버깅 효율성 향상
- 코드 및 문서 전반에 걸쳐 명명 규칙 및 구조 정리

(cherry picked from commit 59cf249c74bf0ca2a985fa259ce1957daeca5329)

pureliture committed:
📝 환경 변수 검증 및 에러 메시지 로깅 추가

- 누락된 환경 변수에 대한 stderr 출력 로깅 추가
- textwrap.dedent를 활용한 CI 환경 변수 가이드 메시지 개선

(cherry picked from commit a6b26c02f4996ef7d91612adfcdb8f0948aa3d08)

pureliture committed:
✨ 로깅 시스템 적용 및 `print` → `logger`로 변경

- 기존의 `print`를 제거하고 로깅 모듈(`logging`)을 사용하도록 변경
- 에러, 경고, 정보 메시지를 적절한 로깅 레벨로 분류
- 스크립트의 주요 구간에 통합 로그 설정 추가
- `.gitignore` 파일 추가로 불필요한 파일 제외 처리

(cherry picked from commit f25ddb0975966f6c6865d8f6c12560862d0376f0)

pureliture committed:
🔒 토큰 암호화/복호화 로직 개선

- 암호화/복호화 코드 가독성 향상 및 오류 디버깅 메시지 추가
- 잘못된 입력 처리 로직 추가로 예외 상황 대비 강화
- 배포 토큰 및 PAT 암호화 로직 정리 및 변수 명시적 정의

ujin2021 commented:
validate_bookmarks.py 파일에서 `필요한 경우 다른 프로젝트에서의 북마크 데이터도 가져와서 통합적으로 검증합니다.` 지금은 다른 플젝에서 갖고오는 건 없고, 확장되었을 때를 고려한 기능일까요?

ujin2021 commented on scripts/fetch_projects.py:
```diff
@@ -0,0 +1,171 @@
+            headers=api_headers
+        )
+
+        if file_response.status_code != 200:
+            logger.warning("⚠️  파일 %s 내용 조회 실패: %s", file_path, file_response.status_code)
+            continue
```
근데 yaml파일 전체를 읽어야 중복검사가 제대로 되잖아요
해당 파일 못읽으면 continue아니고 except 문 써야 하는거 아니에요??

ujin2021 commented on scripts/validate_bookmarks.py:
```diff
@@ -0,0 +1,159 @@
+        auth_method = "PAT" if has_pat else "Deploy Token"
+        logger.info("CI 환경에서 GitLab API 접근 가능 (%s 사용). 모든 프로젝트의 북마크를 검증합니다.", auth_method)
+    else:
+        logger.info("단독 모드로 실행됩니다. 로컬 YAML 파일만 검증합니다.")
```
여기서 검증하는 "로컬 yaml"파일이라 하면 url 생성/수정 하려는 프로젝트 하나만 검증하는거에요??

pureliture committed:
✨ 북마크 검증 로직 개선 및 JSON Schema 도입

- 필수 필드 및 검증 규칙 업데이트
- JSON Schema 기반 검증 로직 추가
- 테스트 스크립트 추가
- Dockerfile 및 README 파일 수정

pureliture committed:
remote stash
